File: BackEnd/PHP/laravelshop/app/Http/Controllers/Api/V1/I18n/I18nController.php
<?php

namespace App\Http\Controllers\Api\V1\I18n;

use App\Http\Controllers\Api\V1\ApiController;
use App\Modules\I18n\Requests\UpsertExchangeRateRequest;
use App\Modules\I18n\Requests\UpsertTranslationRequest;
use App\Modules\I18n\Services\I18nService;
use OpenApi\Attributes as OA;

#[OA\Tag(name: 'I18n')]
class I18nController extends ApiController
{
    public function __construct(
        private readonly I18nService $i18nService,
    ) {}

    public function translations()
    {
        return $this->success($this->i18nService->translations(request()->query('locale'), request()->query('namespace')));
    }

    public function upsertTranslation(UpsertTranslationRequest $request)
    {
        return $this->success($this->i18nService->upsertTranslation($request->user(), $request->validated()));
    }

    public function deleteTranslation(int $id)
    {
        return $this->success($this->i18nService->deleteTranslation(request()->user(), $id));
    }

    public function exchangeRates()
    {
        return $this->success($this->i18nService->exchangeRates());
    }

    public function upsertExchangeRate(UpsertExchangeRateRequest $request)
    {
        return $this->success($this->i18nService->upsertExchangeRate($request->user(), $request->validated()));
    }

    public function convert()
    {
        return $this->success($this->i18nService->convert(
            (float) request()->query('amount'),
            (string) request()->query('from'),
            (string) request()->query('to'),
        ));
    }
}

File: BackEnd/Java/java-spring-maven-jpa-postgresql/src/main/java/com/pbshop/java/spring/maven/jpa/postgresql/i18n/I18nController.java
package com.pbshop.java.spring.maven.jpa.postgresql.i18n;

import java.math.BigDecimal;
import java.util.List;
import java.util.Map;

import jakarta.validation.Valid;

import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import com.pbshop.java.spring.maven.jpa.postgresql.auth.security.AuthenticatedUserPrincipal;
import com.pbshop.java.spring.maven.jpa.postgresql.common.api.ApiResponse;
import com.pbshop.java.spring.maven.jpa.postgresql.i18n.dto.I18nDtos.UpsertExchangeRateRequest;
import com.pbshop.java.spring.maven.jpa.postgresql.i18n.dto.I18nDtos.UpsertTranslationRequest;

@RestController
@RequestMapping("${pbshop.api.base-path:/api/v1}/i18n")
public class I18nController {

    private final I18nService i18nService;

    public I18nController(I18nService i18nService) {
        this.i18nService = i18nService;
    }

    @GetMapping("/translations")
    public ApiResponse<List<Map<String, Object>>> getTranslations(
            @RequestParam(required = false) String locale,
            @RequestParam(required = false) String namespace
    ) {
        return ApiResponse.success(i18nService.getTranslations(locale, namespace));
    }

    @PostMapping("/admin/translations")
    public ApiResponse<Map<String, Object>> upsertTranslation(
            @AuthenticationPrincipal AuthenticatedUserPrincipal principal,
            @Valid @RequestBody UpsertTranslationRequest request
    ) {
        return ApiResponse.success(i18nService.upsertTranslation(principal, request));
    }

    @DeleteMapping("/admin/translations/{id}")
    public ApiResponse<Map<String, Object>> deleteTranslation(
            @AuthenticationPrincipal AuthenticatedUserPrincipal principal,
            @PathVariable Long id
    ) {
        return ApiResponse.success(i18nService.deleteTranslation(principal, id));
    }

    @GetMapping("/exchange-rates")
    public ApiResponse<List<Map<String, Object>>> getExchangeRates() {
        return ApiResponse.success(i18nService.getExchangeRates());
    }

    @PostMapping("/admin/exchange-rates")
    public ApiResponse<Map<String, Object>> upsertExchangeRate(
            @AuthenticationPrincipal AuthenticatedUserPrincipal principal,
            @Valid @RequestBody UpsertExchangeRateRequest request
    ) {
        return ApiResponse.success(i18nService.upsertExchangeRate(principal, request));
    }

    @GetMapping("/convert")
    public ApiResponse<Map<String, Object>> convert(
            @RequestParam BigDecimal amount,
            @RequestParam String from,
            @RequestParam String to
    ) {
        return ApiResponse.success(i18nService.convert(amount, from, to));
    }
}
